[rpc] Add shutdown function to process group agent and refactor join

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30020 [rpc] Add local shutdown to process group agent**

After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `shutdown` method aborts this and joins the all the threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid state. The new way of shutting down is:

```
rpc.wait_all_workers()
rpc.shutdown()
```

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

rohan-varma committed:
Add RpcAgent destructor and stop API

`

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add RpcAgent destructor and stop API"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add RpcAgent destructor and stop API"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -1062,6 +1062,29 @@ def test_call_method_on_rref(self):
 
+    @dist_init(setup_model_parallel=False)
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
+    def test_rpc_agent_destructor(self):
```
Add a small comment here mentioning what we're testing:

"This test ensures that the rpc agent destructor safely shuts down the local RPC framework even if rpc.shutdown is not called"

pritamdamania87 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -123,6 +123,11 @@ PyObject* rpc_init(PyObject* /* unused */) {
     agent->start();
   });
 
+  module.def("local_shutdown_rpc", []() {
```
Lets call this `shutdown` and since this is a public API, lets add appropriate docs explaining what this function does.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -162,6 +162,10 @@ ProcessGroupAgent::ProcessGroupAgent(
 }
 
+ProcessGroupAgent::~ProcessGroupAgent() {
+  localShutdown();
```
This can throw an exception which would be dangerous. Add a try-catch and just LOG(ERROR) the exception here.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -112,6 +112,9 @@ class TORCH_API RpcAgent {
   virtual void start() {}
 
+  // stop accepting requests
+  virtual void localShutdown() {}
```
Call this `shutdown` and also expand the comment: "Stop accepting requests and shutdown the RPC framework as soon as possible by terminating all RPC threads"

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -117,10 +121,16 @@ class ProcessGroupAgent : public RpcAgent {
   std::atomic<int64_t> nextId_;
+  // tracks if the listenerThread_ that processes
+  // incoming messages should be running.
+  std::atomic<bool> start_{false};
```
nit: Call this `rpc_running_`

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -259,12 +263,31 @@ void ProcessGroupAgent::sync() {
+    std::unique_lock<std::mutex> lock(recvWorkMutex_);
+    recvWorkCV_.wait(lock, [this]() { return recvWork_; });
+    recvWork_->abort();
+    start_.store(false);
```
Shouldn't this be set at the beginning?

rohan-varma replied:
If this is set at the beginning, it is possible that the listenerThread would not have started, in which case, `recvWork` would never be valid and we'd wait on the above condition forever. Since we wait on the condition, we'd let the listener thread run at least once and put `recvWork` in a valid state, and then we'd abort it, and prevent the listener thread from running any further

pritamdamania87 replied:
I don't think we should rely on recvWork being in a valid state during shutdown.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -259,12 +263,31 @@ void ProcessGroupAgent::sync() {
+  // CV to make sure we don't abort it in an invalid state.
+  {
+    std::unique_lock<std::mutex> lock(recvWorkMutex_);
+    recvWorkCV_.wait(lock, [this]() { return recvWork_; });
```
What is this wait for?

rohan-varma replied:
This ensures that `recvWork` is in a valid state (i..e. it is not uninitialized, which it would be if there was no call to `recvAnySource` in the `listenerThread`).

xush6528 replied:
It also ensures, for the thread calling stop, it gets the latests `recvWork_` that is being called `wait()`, not the previous RecvWork whose work has completed and whose `wait()` call has been unblocked.

pritamdamania87 replied:
Wouldn't this mean that if we call shutdown without receiving any RPCs, the shutdown will just hang forever? If recvWork_ is not initialized we should just skip the `recvWork_->abort`

> It also ensures, for the user thread calling ::localShutdown(), when it gets the lock, recvWork_ is the latests RecvWork that is being called wait() on by the listenerThread_, rather than the previous RecvWork whose work has completed and whose wait() call has been unblocked.

Doesn't the lock guarantee this? Why do we need the condition_variable wait? It seems like we can get rid of this wait.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -259,12 +263,31 @@ void ProcessGroupAgent::sync() {
+    recvWork_->abort();
+    start_.store(false);
+  }
+  threadPool_.waitWorkComplete();
```
Do we need to do this? Would prefer the local shutdown to be as fast as possible (but safe) and it shouldn't wait on outstanding operations.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -41,6 +41,8 @@ class ProcessGroupAgent : public RpcAgent {
       int numSendRecvThreads,
       std::chrono::milliseconds rpcTimeout);
 
+  virtual ~ProcessGroupAgent() override;
```
Don't think we need virtual here since the base class defines this as virtual.

rohan-varma committed:
Update on "[rpc] Add local shutdown and destructor to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -259,12 +263,31 @@ void ProcessGroupAgent::sync() {
+  }
+  threadPool_.waitWorkComplete();
+  listenerThread_.join();
+  PythonRpcHandler::getInstance().cleanup();
```
We need to call `_destroy_rref_context()` too. If you see `def join_rpc()`.

rohan-varma replied:
I removed this since it was causing RRef leakage issues.

rohan-varma committed:
Update on "[rpc] Add local shutdown and destructor to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown and destructor to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma commented:
Test failure is real, looking into it

rohan-varma commented:
@satgera Would be great if you could take a look at this PR as well

rohan-varma committed:
Update on "[rpc] Add local shutdown and destructor to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown and destructor to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown and destructor to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown and destructor to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -121,6 +121,10 @@ class TORCH_API RpcAgent {
 
+  // Stop accepting requests and shutdown the RPC framework as soon as possible
+  // by terminating all RPC threads.
+  virtual void shutdown() {}
```
make this pure virtual and for the other backend just have a NOOP for now.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -121,6 +121,10 @@ class TORCH_API RpcAgent {
   // start accepting requests
   virtual void start() {}
```
Not necessary for this PR, but this should be pure virtual too.

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -269,14 +271,34 @@ void ProcessGroupAgent::sync() {
   futureTimeoutThread_ =
       std::thread(&ProcessGroupAgent::pollTimedOutRPCs, this);
 }
 
```
What happened to the destructor? I don't see it in the latest version.

rohan-varma replied:
The destructor tests aren't passing, and I'm debugging why. But for now I've removed it and the associated test in this PR, since it's not a strict requirement to be able to shutdown PG agent.

zzzwen commented on test/rpc_test.py:
```diff
@@ -1076,6 +1076,50 @@ def test_call_method_on_rref(self):
+            init_method=self.init_method,
+        )
+        rpc.join_rpc()
+        rpc.shutdown()
```
Question, in this latest version, `join_rpc()` not longer shutdown the agent? And will it block until everyone the rendezvous calls rpc.join_rpc()?

rohan-varma replied:
`join_rpc` will still shutdown everything, the key difference is that it'll also send a message to other workers telling them to shutdown, `localShutdown` won't. Existing workflows shouldn't be affected by this change.

> And will it block until everyone the rendezvous calls rpc.join_rpc()?

If you're referring to whether `rpc.join_rpc` will block until all workers call it, yes, this is true (and not changed by this PR)

rohan-varma replied:
Probably #30050, just because the CI will likely finish first and I'd like to not block that PR any longer (API name changes are especially prone to breakages/merge conflicts the longer they are out there)

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -202,6 +198,10 @@ void ProcessGroupAgent::join() {
       SendWork(allWorkerInfo_[dst], Message({}, {}, MessageType::SHUTDOWN)));
   threadPool_.waitWorkComplete();
   listenerThread_.join();
+  rpcRunning_.store(false);
```
the old `shutdown` variable is now replaced with `rpcRunning`, and the store call is moved here so that the listener thread can process the last shutdown message and shut itself down that way. Then, we need to set this variable and notify the watchdog thread for future timeouts to join as well.

rohan-varma replied:
yes, the difference is that join sends messages to other workers and waits on those messages being processed, while shutdown does not wait on them being processed. Right now both of them terminate the workers, but after a discussion with @satgera @pritamdamania87 and a few others the plan was eventually to be able to call something like.

```
rpc.wait_all_workers() # doesnt shutdown the rpc agent, but synchronizes workers, similar to the old sync_rpc
rpc.shutdown() # does the shutdown
```

pritamdamania87 commented on test/rpc_test.py:
```diff
@@ -1109,11 +1096,6 @@ def test_rpc_join_and_shutdown(self):
-    def test_get_default_rpc_timeout(self):
-        timeout = rpc.get_rpc_timeout()
-        self.assertEqual(timeout, rpc.constants.DEFAULT_RPC_TIMEOUT)
-
```
Why did we remove test_get_default_rpc_timeout?

rohan-varma replied:
This must've show up due to merging/rebasing issues with https://github.com/pytorch/pytorch/pull/30020/files/ - this test was removed there, and we added `test_get_rpc_timeout` instead.

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -269,14 +269,36 @@ void ProcessGroupAgent::sync() {
 }
 
+void ProcessGroupAgent::shutdown() {
+  if (!rpcRunning_.load()) {
```
You're using an atomic, but the check isn't atomic. Multiple threads may call `shutdown` and both of them will execute this function, if they access the atomic's value before one of them has a chance to update it. To make this correct while retaining the atomic (instead of using a mutex every time you access it), you'll have to do do an atomic exchange.

For example:
```c++
// If the value of rpcRunning_ was false already, don't do anything.
if (!rpcRunning_.exchange(false)) {
  return;
}
// Only one thread will have seen the edge from true -> false
```

That said, if you're looking for this behavior AND the guarantee that the agent has shut down when this function returns, check out `std::call_once`. Then, the second thread to call `shutdown` will only return if the first thread has also returned.

rohan-varma replied:
Thanks for this catch. I think the goal is to prevent shutting down twice, as then we'd have issues where we call `join` on threads that have already been `join`ed, for example. I'm also not sure where we could have multiple threads call this shutdown function, as it's only exposed in the python code, so if the user launches multiple threads that all call shutdown, this could be an issue, though that is an unexpected use case.

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -202,6 +198,10 @@ void ProcessGroupAgent::join() {
       SendWork(allWorkerInfo_[dst], Message({}, {}, MessageType::SHUTDOWN)));
   threadPool_.waitWorkComplete();
   listenerThread_.join();
```
is it intentional to join listenerThread_ and cleanup PythonRpcHandler in both join and shutdown?

rohan-varma replied:
Right now, the functionality is in both functions though it will execute only once, since the intended use case is to only call `rpc.shutdown()` or `rpc.wait_all_workers()` for a clean shutdown that waits for other workers.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -164,6 +164,19 @@ Otherwise, throws an exception.
+        agent->shutdown();
+      },
+      R"(
+      Locally shuts down the running RPC agent, without sending messages to
```
I am not sure if this paragraph should start without indent. Please double check if the doc format is expected

rohan-varma replied:
<img width="910" alt="pybind" src="https://user-images.githubusercontent.com/8039770/69399377-c3f3ee00-0ca2-11ea-85e1-9b926d2979b7.png">

This is how the doc looks, indentation looks fine

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -193,16 +193,7 @@ void ProcessGroupAgent::join() {
-  // timeout watchdog would sleep forever.
-  futureTimeoutCV_.notify_one();
-  futureTimeoutThread_.join();
   PythonRpcHandler::getInstance().cleanup();
```
This should not be backend specific and should be moved to a common place. One way would be to have an implementation for `shutdown` in RpcAgent which calls this and then calls a pure virtual method `shutdownImpl`. Each backend can specify their own shutdown mechanism and the common parts are pulled into RpcAgent.

mrshenli replied:
We do call `_cleanup_python_rpc_handler()` in Python `wait_all_workers/join_rpc`. @zhaojuanmao Could you please confirm whether this is still necessary here?

zhaojuanmao replied:
@mrshenli  and @rohan-varma _cleanup_python_rpc_handler() should be only called in python side and it should be called in both wait_all_workers/join_rpc and shutdown() API as these two APIs can be called independently, right? 

if uses write a pure C++ application, there is no python interpreter, we can not clean up in C++ in this case. I guess PythonRpcHandler::getInstance().cleanup() is added accidentally by some diff some time (I did not aware it was added in C++)?

mrshenli replied:
@zhaojuanmao thanks for clarifying. 

@rohan-varma As for now we require applications to call `wait_all_workers()` followed by `shutdown()`, it's probably OK that we just call it in `shutdown()`. But not necessary to trigger tests again due to this. We can discuss in a followup PR about whether we can remove this call in join.

mrshenli commented:
There will be conflicts between this and #30217. I will land #30217 after this PR.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -3,6 +3,7 @@
 #include <c10/util/C++17.h>
 #include <c10d/ProcessGroup.hpp>
 #include <torch/csrc/distributed/rpc/request_callback_impl.h>
+#include <torch/csrc/distributed/rpc/rref_context.h>
```
Why do we need this header here?

rohan-varma replied:
Removed.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -269,14 +260,33 @@ void ProcessGroupAgent::sync() {
+  listenerThread_.join();
+  futureTimeoutCV_.notify_one();
+  futureTimeoutThread_.join();
+  // PythonRpcHandler::getInstance().cleanup();
```
Any reason for doing cleanup in `join` instead of here?

rohan-varma replied:
This wasn't right, I moved it to `shutdown()`.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -186,23 +186,14 @@ void ProcessGroupAgent::join() {
   //    feed it a message or kill the thread.
   // 2. A GLOO process cannot send message to itself. (there is an ongoing
   //    effort to fix this problem).
```
The above comments are no longer relevant, let's remove.

xush6528 commented on test/rpc_test.py:
```diff
@@ -1091,6 +1105,49 @@ def test_call_method_on_rref(self):
+
+    @dist_init(setup_rpc=False)
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
+    def test_rpc_join_and_shutdown(self):
```
test method name?

test_ wait_all_workers_and_shutdown

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -186,23 +186,14 @@ void ProcessGroupAgent::join() {
-  int dst = (pg_->getRank() + 1) % pg_->getSize();
-  enqueueSend(
-      SendWork(allWorkerInfo_[dst], Message({}, {}, MessageType::SHUTDOWN)));
   threadPool_.waitWorkComplete();
```
We can keep it as it for now, but we probably don't need this call any more. The original intention is to wait for the `SHUTDOWN` message to be sent, but now we no longer need that message.

rohan-varma replied:
This is removed in the latest iteration

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -456,10 +466,22 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+    if (!rpcRunning_.load()) {
+      return;
+    }
+    bool aborted = !recvWork_->wait();
```
Question: sorry I didn't closely follow the abort work. Does the order of `recvWork_->abort();` and `recvWork_->wait()` matter?

rohan-varma replied:
In this case, the order should be such as:
T1: Calls into `recvWork->wait()`
T2: Aborts the above with `recvWork->abort()`
T1: returns out of recvWork, finishes thread

mrshenli replied:
What if `abort()` is called before T1 calls `wait()`? Will the `wait()` immediately returns with `aborted = true`?

rohan-varma replied:
Yep, this code eventually calls this in gloo: https://github.com/facebookincubator/gloo/blob/master/gloo/transport/tcp/unbound_buffer.cc#L61-L63

that block would translate to something like
`while(!abortWaitRecv_ && recvCompletions <= 0) {
    wait(lock, timeout);
}`

so if `abort()` is called before T1 calls `wait`, then when T1 calls into wait it will return immediately.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -456,10 +466,22 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
 
 void ProcessGroupAgent::listenLoop() {
-  while (true) {
+  while (rpcRunning_.load()) {
```
Why do we need to check `rpcRunning_.load()` twice in each iteration?  Will it cause any problem if we do `while(true)` here? Or is it just to save the overhead of creating the preamble?

mrshenli commented on test/rpc_test.py:
```diff
@@ -1091,6 +1105,49 @@ def test_call_method_on_rref(self):
+        # too early.
+        rpc.rpc_sync("worker{}".format(dst_rank), _set_rpc_done, args=(1,))
+        _check_rpc_done(1)
+        rpc.shutdown()
```
Is this intentional to call `shutdown` without `wait_all_workers`?

rohan-varma replied:
Yes, we'd like to test ability to do a local shutdown without having to wait for outstanding work to be completed.

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

zhaojuanmao commented:
@rohan-varma  please see comments regarding _cleanup_python_rpc_handler. also I'm wondering whether we can rename shutdown to local_shutdown?

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma commented:
> also I'm wondering whether we can rename shutdown to local_shutdown?

I think this was the original name, but @pritamdamania87 and a few others suggested to call it `shutdown`.

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Thanks for adding this!

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] Add local shutdown to process group agent"


After changes in the previous PR (#29928) and the gloo code (https://github.com/facebookincubator/gloo/pull/232), we can now interrupt the listener thread in process group agent, which was blocked on `wait`ing for some work to be received. The `localShutdown` method aborts this thread and joins the other threads. A condition variable is also added to ensure that we don't abort the thread while `recvWork` is in an invalid statem.

A destructor is also added which simply calls `localShutdown`, and we ensure that this function is not called multiple  times.

Differential Revision: [D5578006](https://our.internmc.facebook.com/intern/diff/D5578006/)

[ghstack-poisoned]

rohan-varma commented on test/rpc_test.py:
```diff
@@ -1091,6 +1105,50 @@ def test_call_method_on_rref(self):
+        # too early.
+        rpc.rpc_sync("worker{}".format(dst_rank), _set_rpc_done, args=(1,))
+        _check_rpc_done(1)
+        rpc.wait_all_workers()
```
Seeing occasional flakiness for this test, so `wait_all_workers` needs to be added here.

pietern reviewed (commented):
Copying comment I made inline somewhere:

> So shutdown is a subset of join?
> 
> To me, shutdown implies it being graceful, which is already the job of join. I think terminate or kill is more appropriate here, since there is no expectation that workers will complete outstanding work.

If I see `shutdown` I'm assuming some form of gracefulness. I think `terminate` is more appropriate here, because as implemented, it's equivalent to pulling the plug. If you want to stick to `shutdown`, that's fine with me as well, but please add a couple of lines to the documentation stating that this is NOT not a graceful shutdown.

rohan-varma commented:
@pietern - Sorry I missed this comment. I agree that if the `shutdown()` function is called by itself here, it's a non-graceful shutdown. Though the intent is for this function to be called after `wait_all_workers()`, in which case the shutdown will be graceful. I will send a follow up PR to mention in the docs that this is not graceful when called without `wait_all_workers`

zhaojuanmao commented:
if this shutdown() has dependency on wait_all_workers(), that being said, wait_all_workers() and shutdown() has to be called together. why do we need to split them?

mrshenli commented:
This PR breaks master, reverting:

https://app.circleci.com/jobs/github/pytorch/pytorch/3700145

```
Nov 22 18:54:15 ======================================================================
Nov 22 18:54:15 FAIL [1.053s]: test_wait_all_workers_and_shutdown (__main__.RpcTestWithSpawn)
Nov 22 18:54:15 ----------------------------------------------------------------------
Nov 22 18:54:15 Traceback (most recent call last):
Nov 22 18:54:15   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 130, in wrapper
Nov 22 18:54:15     self._join_processes(fn)
Nov 22 18:54:15   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 211, in _join_processes
Nov 22 18:54:15     self._check_return_codes(elapsed_time)
Nov 22 18:54:15   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 235, in _check_return_codes
Nov 22 18:54:15     self.assertEqual(first_process.exitcode, 0)
Nov 22 18:54:15   File "/var/lib/jenkins/workspace/test/common_utils.py", line 839, in assertEqual
Nov 22 18:54:15     super(TestCase, self).assertLessEqual(abs(x - y), prec, message)
Nov 22 18:54:15 AssertionError: 1 not less than or equal to 1e-05 : 
```
